src/deadline/unreal_submitter/unreal_dependency_collector.py

```diff
@@ -0,0 +1,182 @@
+    """
+    Class container that contains methods for filtering dependencies
+    """
+
+    @staticmethod
+    def dependency_in_game_folder(dependency_path: str) -> bool:
+        """Checks if /Game/ exists in passed path
+
+        :param dependency_path: The path to dependency
+        :return: True if dependency path in `/Game/`, False otherwise
+        :rtype: bool
+        """
+        return "/Game/" in str(dependency_path)
+
+
+@dataclass
+class DependencyOptions:
+    """
+    Dataclass which contains the default options for getting the asset dependencies
+    """
+
+    include_hard_package_references: bool = True
+    include_soft_package_references: bool = True
+    include_hard_management_references: bool = True
+    include_soft_management_references: bool = True
+    include_searchable_names: bool = False
+
+    def as_dict(self):
+        """
+        Return Dependency Search Options as dictionary
+
+        :return: search options dictionary
+        :rtype: dict
+        """
+
+        return asdict(self)
+
+
+class DependencyCollector:
+
+    def __init__(self):
+        self._all_dependencies = list()
+        self._missing_dependencies = list()
+        self._already_synced = list()
+
+    def collect(
+        self,
+        asset_path: str,
+        dependency_options=DependencyOptions(),
+        filter_method: Callable = None,
+        sync_missing: bool = True,
+    ) -> list[str]:
+        """The method starts the algorithm for obtaining dependencies on the passed asset
+
+        :param asset_path: Path to the dependency asset to be obtained
+        :param dependency_options: Settings for obtaining dependencies.
+        :param filter_method: Method by which dependencies will be filtered for synchronization
+        :param sync_missing: Whether to sync missing dependencies or not
+
+        :return: list of collected dependencies (UE paths, start with /Game, etc.)
+        :rtype: list[str]
+        """
+
+        self._all_dependencies.clear()
+
+        udependency_options = unreal.AssetRegistryDependencyOptions(**dependency_options.as_dict())
+
+        if not unreal.EditorAssetLibrary.does_asset_exist(asset_path):
+            self._sync_assets([asset_path])
+
+        _, missing_dependencies = self._get_dependencies(
+            asset_path, udependency_options, filter_method
+        )
+        while missing_dependencies:
+            if sync_missing:
+                self._sync_assets(missing_dependencies)
+            # Anyway extend synced even if source control is not available.
+            # In that case first recursion before loop and recursion in loop will run one time
+            self._already_synced.extend(missing_dependencies)
+            self._missing_dependencies.clear()
+            self._all_dependencies.clear()
+            _, missing_dependencies = self._get_dependencies(
+                asset_path, udependency_options, filter_method
+            )
+
+        self._sync_assets(
```

Comment:
We this option need to be sure that user send the up-to-date files. But since we decide relies on current changelist number, `sync_missing` flag set as False in the `collect()` calls.

Reply:
Not quite sure I follow - this is pulling files down from P4 to the client workspace on the submitter side, but why would that matter?  If we're using a changelist number on the adapter side, then as long as that side syncs to that changelist before rendering everything would be up to date?

Reply:
Yes, I think we shouldn't sync it here. If the user doesn't have soft references, this sync won't help. Additionally, this sync might change the user's workspace, which they may prefer to avoid. So, @baxeaz  is right -- we should sync only on the adaptor side